algopy_testing/tests/primitives/test_biguint.py

```diff
@@ -0,0 +1,740 @@
+import operator
+import re
+from collections.abc import Callable
+
+import algokit_utils
+import pytest
+from algopy import BigUInt, UInt64
+from algopy.constants import MAX_UINT64, MAX_UINT512
+
+from tests.primitives.conftest import AVMInvoker
+from tests.primitives.util import int_to_bytes
+
+_negative_value_error = "expected positive value"
+_too_big512_error = re.escape(f"expected value <= {MAX_UINT512}")
+
+_underflow_error = "- underflows"
+
+_avm_overflow_error = "math attempted on large byte-array"
+_avm_underflow_error = "byte math would have negative result"
+_avm_zero_division_error = "division by zero"
+_avm_zero_mod_error = "modulo by zero"
+
+
+def _object_as_biguint(x: object) -> BigUInt | None:
+    return BigUInt.from_bytes(x) if (isinstance(x, bytes)) else None
+
+
+def _get_avm_result(
+    get_avm_result: AVMInvoker, method: str, a: int, b: int | UInt64
+) -> tuple[BigUInt | None, object]:
+    a_bytes = int_to_bytes(a)
+    if isinstance(b, UInt64):
+        avm_result = get_avm_result(method, a=a_bytes, b=b.value)
+    else:
+        b_bytes = int_to_bytes(b)
+        avm_result = get_avm_result(method, a=a_bytes, b=b_bytes)
+    return (_object_as_biguint(avm_result), avm_result)
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        0,
+        1,
+        MAX_UINT512,
+    ],
+)
+def test_biguint_unary_plus(value: int) -> None:
+    # just test the implementation as there is no AVM equivalent
+    assert +BigUInt(value) == value
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        0,
+        1,
+        MAX_UINT512,
+    ],
+)
+def test_biguint_bool(value: int) -> None:
+    # just test the implementation as there is no AVM equivalent
+    assert bool(BigUInt(value)) == bool(value)
+
+
+@pytest.mark.parametrize(
+    ("a", "b"),
+    [
+        (0, 0),
+        (0, 1),
+        (0, MAX_UINT512),
+        (1, 0),
+        (1, 1),
+        (1, MAX_UINT512),
+        (MAX_UINT512, MAX_UINT512),
+    ],
+)
+@pytest.mark.parametrize(
+    "op_name",
+    [
+        "eq",
+        "ne",
+        "lt",
+        "le",
+        "gt",
+        "ge",
+    ],
+)
+def test_biguint_comparison(get_avm_result: AVMInvoker, op_name: str, a: int, b: int) -> None:
```

Comment:
Since having a BigUInt with a value > MAX_UINT512 is possible, we should add tests to cover these cases for the comparison operations